Data/Pool.hs

```diff
@@ -274,35 +275,64 @@ withResource pool act = control $ \runInIO -> mask $ \restore -> do
 -- This function returns both a resource and the @LocalPool@ it came from so
 -- that it may either be destroyed (via 'destroyResource') or returned to the
 -- pool (via 'putResource').
-takeResource :: Pool a -> IO (a, LocalPool a)
-takeResource pool@Pool{..} = do
-  local@LocalPool{..} <- getLocalPool pool
-  resource <- liftBase . join . atomically $ do
-    ents <- readTVar entries
-    case ents of
-      (Entry{..}:es) -> writeTVar entries es >> return (return entry)
-      [] -> do
-        used <- readTVar inUse
-        when (used == maxResources) retry
-        writeTVar inUse $! used + 1
-        return $
-          create `onException` atomically (modifyTVar_ inUse (subtract 1))
-  return (resource, local)
-#if __GLASGOW_HASKELL__ >= 700
+--
+-- This function returns 'Nothing' if it a resource is not acquired in
+-- the alotted time (timeout).
+takeResource :: Pool a -> IO (Maybe (a, LocalPool a))
+takeResource pool@Pool{..} =
+  case timeout of
+    Just poolTimeout -> takeResourceWithTimeout poolTimeout pool
+    Nothing -> do
+      local@LocalPool{..} <- getLocalPool pool
+      resource <- liftBase . join . atomically $ do
+        ents <- readTVar entries
+        case ents of
+          (Entry{..}:es) -> writeTVar entries es >> return (return entry)
+          [] -> do
+            used <- readTVar inUse
+            when (used == maxResources) retry
+            writeTVar inUse $! used + 1
+            return $ create `onException` atomically (modifyTVar_ inUse (subtract 1))
+      return $ Just (resource, local)
```

Comment:
I think we should make this something like...
```
takeResource pool@Pool{..} =
  let mbTimeout = fromMaybe maxBound 
   in ... <inline body of takeResourceWithTimeout using mbTimeout> ...
```
This will be DRY, and won't have ambiguity/weirdness of passing both `poolTimeout` and `Pool{timeout}` to the same function. The downside is we register a pointless timeout in the case where we need to create a new connection, but I think that's a fine tradeoff.
What do you think?

Reply:
Interesting. Maybe this ^ but we also start off with a call to `tryTakeResource` first. If it's immediate, we don't pay any extra price. If it's not, we just go ahead and register the timeout.